Add acceptance tests for `repo` commands

jtmcg opened the pull request:
Closes https://github.com/github/cli/issues/589

## Changes

Implements a suite of acceptance tests for the `gh repo` command, including coverage of the following sub-commands:

- archive

- clone
- create
- delete
- deploy-key
  - add
  - delete
  - list
- edit
- fork
- list
- rename
- set-default
- sync
- unarchive
- view

Some of these commands are tested in the same files, such as `gh repo fork` and `gh repo sync` in the `repo-fork-sync.txtar` file, as the testing for one command (`gh repo sync`, in this example) is either enabled or made easier by the use of the other.

## Question/Concern

The way that testing for `gh repo deploy-key add` is set up now will result in a new deploy key being set up on behalf of whoever runs this test. That results in an email to the user saying a new deploy key was added. 

I doubt there's any security risk, here, as the private key has been deleted and will remain unused and the repo gets deleted after the test is running, meaning the public key is now useless but for its format for testing purposes.

Ultimately, however, we'll probably want to change the authentication of these tests to a bot user - especially if/when these become part of our CI pipeline - and that will eliminate the emails. It's probably fine for now, as the only way we're going to run these tests is locally, but I wanted to call it out here.

What am I missing? What concerns with the current approach do y'all have? 

## Testing

1. Pull down branch: `gh pr checkout 9783`
2. Run the acceptance tests in the repo directory: 

```shell
GH_ACCEPTANCE_HOST=<HOST> \
GH_ACCEPTANCE_ORG=<ORG> \
GH_ACCEPTANCE_TOKEN=<TOKEN> \
go test -tags=acceptance -run ^TestRepos$ ./acceptance
```

jtmcg committed:
Added test function for repos and repo-create test

jtmcg committed:
Added acceptance test for repo-delete

jtmcg committed:
Add acceptance test for repo-clone

jtmcg committed:
Acceptance testing for repo-archive and repo-unarchive

jtmcg committed:
Add acceptance tests for repo-list and repo-rename

jtmcg committed:
Add acceptance test for repo-edit

jtmcg committed:
Add acceptance test for repo-set-default

jtmcg committed:
Add acceptance tests for repo-fork and repo-sync

jtmcg committed:
Add acceptance tests for repo deploy-key add/list/delete

jtmcg commented on acceptance/testdata/repos/repo-edit.txtar:
```diff
@@ -0,0 +1,19 @@
+
+# Check that the repo description is empty
+exec gh repo view $ORG/$SCRIPT_NAME-$RANDOM_STRING --json description --jq '.description'
+stdout ''
```
This will always be true, as it matches regex. Need a wildcard, instead:

```suggestion
stdout '.'
```

jtmcg commented on acceptance/testdata/repos/repo-create-view.txtar:
```diff
@@ -0,0 +1,12 @@
+defer gh repo delete --yes $ORG/$SCRIPT_NAME-$RANDOM_STRING
+
+# Check that the repo exists
+exec gh repo view $ORG/$SCRIPT_NAME-$RANDOM_STRING --json name --jq '.name'
```
Let's make this match the other tests here with `=` in the flags

```suggestion
exec gh repo view $ORG/$SCRIPT_NAME-$RANDOM_STRING --json=name --jq='.name'
```

jtmcg commented on acceptance/testdata/repos/repo-list-rename.txtar:
```diff
@@ -0,0 +1,24 @@
+exec gh repo create $ORG/$SCRIPT_NAME-$RANDOM_STRING --add-readme --private
+
+# List the repos and check for the new repo
+exec gh repo list $ORG --json name --jq '.[].name'
```
Change to `=` in flags

```suggestion
exec gh repo list $ORG --json=name --jq='.[].name'
```

jtmcg commented on acceptance/testdata/repos/repo-list-rename.txtar:
```diff
@@ -0,0 +1,24 @@
+exec gh repo list $ORG --json name --jq '.[].name'
+stdout $SCRIPT_NAME-$RANDOM_STRING-renamed
+
+# Delete the renamed repo
+exec gh repo delete $ORG/$SCRIPT_NAME-$RANDOM_STRING-renamed --yes
+
+# List the repos and check for the deleted repo
+exec gh repo list $ORG --json name --jq '.[].name'
+! stdout $SCRIPT_NAME-$RANDOM_STRING-renamed
+
```
Don't need this, here. Let's just defer delete above

jtmcg commented on acceptance/testdata/repos/repo-set-default.txtar:
```diff
@@ -0,0 +1,23 @@
+# Defer repo cleanup
+defer gh repo delete --yes $ORG/$SCRIPT_NAME-$RANDOM_STRING
+
+# Clone the repo
+exec gh repo clone $ORG/$SCRIPT_NAME-$RANDOM_STRING
```
Let's collapse into the create command

jtmcg committed:
Cleanup some inconsistencies and improve collapse some functionality

jtmcg commented on acceptance/testdata/repos/repo-archive-unarchive.txtar:
```diff
@@ -0,0 +1,26 @@
+# Defer repo cleanup
+defer gh repo delete --yes $ORG/$SCRIPT_NAME-$RANDOM_STRING
+
+# Check that the repo exists and isn't archived
+exec gh repo view $ORG/$SCRIPT_NAME-$RANDOM_STRING --json=isArchived --jq='.isArchived'
+stdout false
```
Cross referencing a [conversation](https://github.com/cli/cli/pull/9782#discussion_r1805268802) from @andyfeller's PR:

Paraphrasing @williammartin comment:

> Can you talk to me a bit about why you feel you need to check a [precondition] before proceeding? How's this moving the needle on your confidence levels?

I think for me it was about instilling confidence that the thing I'm going to assert against to define success of a command was indeed changed or affected. It was the loop I did myself in the terminal to gain confidence that the commands were working as expected.

There is an additional effect in that it protects against unintended changes to a command's default behavior. In my example, if, for some reason, `gh repo create` started automatically creating `archived` repos, we'd catch it here. Maybe that's out of scope of my test? 🤔

williammartin replied:
Thanks for the explanation.

I think that "hey script reader, up front here's the thing this specific test cares about changing and I'm communicating that to you" is probably the thing that resonates with me most. That's why in the other PR I suggested that if we did have these precondition checks that they are the same assertion differing only in whether they are expecting or refuting state.

I think that if we wanted to validate that repos are created in an unarchived form, then that should be tested against `repo create` directly, so that the confidence in that command is dependent on some other command that may or may not be testing it.

But I want to be clear here, I'm **not** saying that these should be removed. I am saying that I think it's important that there is a clear and shared pattern to all the testscripts that we can say "make it look like that". That can also mean going back to the scripts that are already being merged and making them use this pre and post pattern.

jtmcg replied:
> "hey script reader, up front here's the thing this specific test cares about changing and I'm communicating that to you"

This is definitely my primary motivation for doing it. I hear you on the pattern. That's probably a good chat to have synchronously. Maybe we talk about it in retro?

williammartin commented on acceptance/testdata/repos/repo-fork-sync.txtar:
```diff
@@ -0,0 +1,44 @@
+# Use gh as a credential helper
+exec gh auth setup-git
```
**Change Requested**

I believe this is the only test that actually needs `gh` configured as a credential helper. Can you remove it from the rest?

williammartin commented on acceptance/testdata/repos/repo-archive-unarchive.txtar:
```diff
@@ -0,0 +1,26 @@
+
+# Check that the repo exists and isn't archived
+exec gh repo view $ORG/$SCRIPT_NAME-$RANDOM_STRING --json=isArchived --jq='.isArchived'
+stdout false
```
**Suggestion**

I think it would be nice for all `stdout` assertions where possible (I know you can't wrap env vars up) to be wrapped in single quotes to make it clear we're dealing with string comparison here. It might just be my sensitivity but booleans on their own seem to indicate some other meaning.

williammartin commented on acceptance/testdata/repos/repo-clone.txtar:
```diff
@@ -0,0 +1,15 @@
+exec gh repo clone $ORG/$SCRIPT_NAME-$RANDOM_STRING
+
+# Ensure the repo was cloned
+exec ls $SCRIPT_NAME-$RANDOM_STRING
```
**Change Requested**

I think the testscript idiomatic way to do this is:

```
exists $SCRIPT_NAME-$RANDOM_STRING/README.md
```

It's more portable by not relying on `ls` existing on the PATH.

williammartin commented on acceptance/testdata/repos/repo-delete.txtar:
```diff
@@ -0,0 +1,16 @@
+exec gh repo delete --yes $ORG/$SCRIPT_NAME-$RANDOM_STRING
+
+# Ensure that the repo was deleted
+! exec gh repo view $ORG/$SCRIPT_NAME-$RANDOM_STRING --json name --jq '.name'
```
**Nit**

Could go either way on this but since we expect this to error the `--json ...` onwards is kind of misleading.

jtmcg replied:
I think that the thought here was that the negation matched the assertion from before the operation, but I'm not sure it matters... However, I generally think less is more, so I'll make this change

williammartin commented on acceptance/testdata/repos/repo-deploy-key.txtar:
```diff
@@ -0,0 +1,34 @@
+
+# Ensure the deploy key was deleted
+exec gh repo deploy-key list --json=id --jq='.[].id'
+! stdout $DEPLOY_KEY_ID
```
Very clever incantations 👍

williammartin commented on acceptance/testdata/repos/repo-fork-sync.txtar:
```diff
@@ -0,0 +1,44 @@
+defer gh repo delete $ORG/$SCRIPT_NAME-$RANDOM_STRING-fork --yes
+
+# Sleep so that the BE has time to sync
+sleep 5
```
**Question**

I was noodling on something with @BagToad yesterday that I want to throw at you. How would you feel if instead of:

```
# Sleep so that the BE has time to sync
sleep 5

# Check that the repo was forked
exec gh repo view $ORG/$SCRIPT_NAME-$RANDOM_STRING-fork --json='isFork' --jq='.isFork'
stdout true
```

We had something along the lines of:

```
# Check that the repo was forked
retryableexec gh repo view $ORG/$SCRIPT_NAME-$RANDOM_STRING-fork --json='isFork' --jq='.isFork'
eventuallystdout true
```

Where the command would be repeated on loop every `n` milliseconds up to `m` times.

Working example seen here:

https://github.com/cli/cli/commit/f98f45fce9e9c34abb231efe9ba9b821ca684315#diff-0baffab94029cd95fd0c6237468ce85738db4ea07d63397348367930b7ad8833

jtmcg replied:
I would absolutely love that! I had briefly thought about some sort of polling solution in place of `sleep`, but thought that since `sleep` got the job done, here, I'd punt it until later. As such, this is right in line with what I would expect, here 👍

williammartin replied:
Yes totally 100% let's keep `sleep`. There are some edge cases to figure out around the other one, I just wanted to float it to get your opinion.

williammartin commented on acceptance/testdata/repos/repo-fork-sync.txtar:
```diff
@@ -0,0 +1,44 @@
+exec gh repo sync
+
+# Check that asset.txt now exists in the fork
+exec ls
```
**Change Requested**

As above I think `exists` here would be better.

williammartin reviewed (changes requested):
Lots of good stuff in here, reads well to me. As mentioned in the comment, I can see the value of the pre-condition post-condition check when the commands are duals / mirrors of each other.

jtmcg committed:
Remove unnecessary gh auth setup-git steps

jtmcg committed:
Wrap boolean strings in '' so it is clear they are strings

jtmcg committed:
Replace LICENSE
Makefile
README.md
acceptance
api
bin
build
cmd
context
docs
git
go.mod
go.sum
internal
pkg
script
share
test
utils commands with

jtmcg committed:
Remove unnecessary flags from repo-delete testscript

williammartin commented on acceptance/acceptance_test.go:
```diff
@@ -72,6 +72,14 @@ func TestReleases(t *testing.T) {
+	if err := tsEnv.fromEnv(); err != nil {
+		t.Fatal(err)
+	}
+	testscript.Run(t, testScriptParamsFor(tsEnv, "repos"))
```
Oh sorry, one more minor thing for consistencies sake, we've been naming the dir after the command i.e. `repo`.

jtmcg replied:
Gah, you're right... Total oversight on my part 🤦 Definitely gonna make this change

williammartin reviewed (approved):
Approving with the understanding that you address the last review comment (either by doing it or rejecting it).

```
➜  cli git:(jtmcg/testscripts-repo) set -o pipefail && GH_ACCEPTANCE_HOST=github.com GH_ACCEPTANCE_ORG=gh-acceptance-testing go test -tags acceptance -json -coverprofile=coverage.out -coverpkg=./... -run ^TestRepos$ github.com/cli/cli/v2/acceptance | tparse --all go test
┌──────────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │               TEST               │             PACKAGE               │
│─────────┼─────────┼──────────────────────────────────┼───────────────────────────────────│
│  PASS   │   18.09 │ TestRepos/repo-fork-sync         │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    5.99 │ TestRepos/repo-deploy-key        │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    5.93 │ TestRepos/repo-archive-unarchive │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    4.47 │ TestRepos/repo-list-rename       │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    4.23 │ TestRepos/repo-edit              │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    4.09 │ TestRepos/repo-set-default       │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    3.62 │ TestRepos/repo-clone             │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    3.50 │ TestRepos/repo-create-view       │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    3.38 │ TestRepos/repo-delete            │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    0.00 │ TestRepos                        │ github.com/cli/cli/v2/acceptance  │
└──────────────────────────────────────────────────────────────────────────────────────────┘
┌────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │             PACKAGE              │ COVER │ PASS │ FAIL │ SKIP  │
│─────────┼─────────┼──────────────────────────────────┼───────┼──────┼──────┼───────│
│  PASS   │ 19.14s  │ github.com/cli/cli/v2/acceptance │ 12.4% │  10  │  0   │  0    │
└────────────────────────────────────────────────────────────────────────────────────┘
```

jtmcg committed:
Rename acceptance test directory from repos to repo

jtmcg committed:
Wrap true in '' in repo-fork-sync

jtmcg commented:
Force pushes to sync with `trunk` and clean up history
